### ISSUE TITLE
Choices for modes, reproducing results and using NYT dataset

### ISSUE THREAD

**Issue by soumyac1999** [morningmoni/HiLAP#6]
__Question 1__
In `conf.py`, the given options are `'hilap', 'sl', 'hilap-sl', 'hmcn'`.  Could you please explain what the `sl` option is?

__Question 2__
What configuration options should I use to reproduce the results for (a) _TextCNN (Flat)_ and (b) _HiLAP(TextCNN)_ given in Table 2 of the paper? 

__Question 3__
In issue #3, you had said that first `step-sl` should be used. I think this means that first `hilap-sl` and the model obtained should be further trained using `hilap`, is this correct?

I tried `hilap-sl` followed by `hilap` on RCV1 dataset but the results don't seem to match the ones in the paper. Specifically, the used command and obtained results are:

```bash
# First hilap-sl
py main.py --mode hilap-sl
# Result -> best: (0.7686, 0.3392, 50), (0.7686, 0.3392, 50)

# Followed by hilap
py main.py --mode hilap --load_model runs/<run_date_and_time>_rcv1_textcnn_hilap-sl_del/epoch50_0.6052_0.7454_0.6067.pth.tar --isTrain True
# Result -> best: (0.7849, 0.4874, 20), (0.7797, 0.4939, 40)

# Results in paper -> HiLAP (TextCNN) 78.6 50.5
```
What else needs to be done to reproduce the reported results?

__Question 4__
In `readData_nyt.py`, could you please specific how the files `filelist_` (Line 16, 28), `accum` (L 32), etc. are generated?

Thanks!

**Comment by ayushbits** [comment]
@morningmoni 
i've question wrt Q2 in the above post.

> Question 2
> What configuration options should I use to reproduce the results for (a) TextCNN (Flat) and (b) HiLAP(TextCNN) given in Table 2 of the paper?

What config options should I choose to reproduce part  TextCNN (Flat) numbers ?

Please help !
Thanks

**Comment by morningmoni** [comment]
Thank you both for your interest.

Q1.
sl is short for supervised learning (the flat model).

Q2 & Q3
Yes (I just changed the names to those in the papers to avoid confusion)

From your results, it looks like the Micro-F1 scores are close but the Macro-F1 scores, esp for HiLAP-SL are quite off. Could you double-check HiLAP-SL first to make sure it is close to the results in the paper (Fig. 4) and then train HiLAP?

For config options, most of them are supposed to work fine without changing. We didn't do extensive hyper-parameter search for generalizability. The hidden sizes are different on the datasets with raw inputs (FUN/GO, already written in the `conf.py`). For debugging, You can disable the mixture of different modes (set`sl_ratio` that combines HiLAP and HiLAP-SL losses, and `global_ratio` that combines Flat and HiLAP-SL losses to 0) and see how the performance would change.

Q4
Those files are not generated but the original files in the NYT dataset. You can find them following the link I believe (needs license unfortunately).

**Comment by soumyac1999** [comment]
Thanks for you reply @morningmoni.

> sl is short for supervised learning (the flat model)

I ran the flat models using the mode `sl` and the results which I obtained are way off from the reported ones and appear counter-intuitive since they performance appears to be better than the hierarchical approaches. What could have gone wrong and what should I be doing for reproducing the flat results?

```bash
py main.py --mode sl
# Result -> best: (0.8021, 0.5087, 0), (0.8021, 0.5087, 0)

# Outputs in case they help with debugging (I have omitted date and time)
# main.py[line:590][evaluate] INFO-> acc:0.8685 f1_s:0.8022 micro-f1:0.7962 0.7962 0.7986 macro-f1:0.5033 0.5033 0.5041
# main.py[line:595][evaluate] WARNING-> best: (0.8021, 0.5087, 0), (0.8021, 0.5087, 0)

# Another run -> best: (0.8019, 0.4964, 0), (0.8005, 0.511, 0)
```

> Could you double-check HiLAP-SL first to make sure it is close to the results in the paper (Fig. 4) and then train HiLAP?

I ran `py main.py --mode hilap-sl` two more times and the results are
```
Run1 -> best: (0.7856, 0.4693, 20), (0.778, 0.4904, 40)
Run2 -> best: (0.7861, 0.4744, 20), (0.7796, 0.4887, 40)
```
I think these are close to ones reported in the paper. Since the results don't seem to be deterministic, how are the numbers in the paper calculated? (Are they averaged over multiple runs or something else?) Also, how are the reported Micro and Macro F1's in Table 2 calculated from the two tuples that are printed (The tuples correspond to epochs for best Micro and Macro, right?)

> Those files are not generated but the original files in the NYT dataset. You can find them following the link I believe (needs license unfortunately).

I could not find those files in the NYT dataset. Only directories for various years are present in the `data` directory. Had some preprocessing been done on the dataset to generate the other files? 
```
$ ls data
1987  1989  1991  1993  1995  1997  1999  2001  2003  2005  2007
1988  1990  1992  1994  1996  1998  2000  2002  2004  2006
```

**Comment by ayushbits** [comment]
Hi @morningmoni ,
Thanks for the reply. As @soumyac1999 mentioned, even I tried with the same config parameters and results seems different than paper.
In the RCV1 dataset, it has same number of training and test as mentioned in your paper.
Used glove 6B.50d

`python main.py --mode sl
best: (0.8014, 0.4935, 0), (0.8007, 0.5082, 0)`

**Comment by morningmoni** [comment]
Hi both,

This is interesting. I just tried running flat TextCNN twice on my old codebase and failed to achieve the performance you reported... It might be that I broke (improved?) something during code refactoring (quite a while ago). I can take a closer look to see what's going on later. Also, the results in the paper are already much better than some previously reported numbers (Table 8). 
On the other hand, pls note that the tuples show the best performance which is only used for reference, and the numbers in the paper are (supposed to be) lower. We ran each mode for the same number of epochs and reported their average performance at the same epoch. 
Meanwhile, if I did improve the base model somehow, you may also try tuning the 'sl-ratio' and 'global-ratio' as I mentioned to see if you can get an even better 'hybrid' system.

@soumyac1999, for the NYT dataset, I double-checked and assumed that "filelist_1987-02.txt" and "filelist_2003-07.txt" contain filenames divided by year (1987-2002, 2003-2007) per line, and accum_* are the two folders storing the files. Sorry I was not the person who processed the data.

➜  data head -3 filelist_1987-02.txt
0009003.xml
0009004.xml
0009005.xml

**Comment by ayushbits** [comment]
Thanks for the reply @morningmoni 
Can you share the older code base from which you are getting paper numbers (for Textcnn - Flat) ?
We will also compare the older and current code.

Thanks

**Comment by ayushbits** [comment]
Hi @morningmoni ,
Can you provide the older code base (TextCnn- flat) that was producing paper's numbers ?

**Comment by morningmoni** [comment]
Can you run it again using this [data file](https://drive.google.com/file/d/1hTC7SSxMZ7IwR1NOLOxBiQ-kCBSj9z1m/view?usp=sharing) and see what the performance is? I suspect there is some subtle difference in the raw data (since it wasn't processed by myself).

**Comment by soumyac1999** [comment]
I ran `sl` with your pickle dataset and the results are now closer to the ones reported.
```
Run1: best: (0.7826, 0.4641, 0), (0.7826, 0.4641, 0)
Run2: best: (0.7808, 0.4557, 0), (0.7808, 0.4557, 0)
```

Can you please let me know what additional preprocessing has been done on the data since I thought the `readData_rcv1.py` directly uses the raw dataset?

Thanks!

**Comment by morningmoni** [comment]
That's also what I thought... I don't know if there is any additional preprocessing on the raw dataset. Perhaps there is some issue with the raw dataset I have. My suggestion is: if you are trying to reproduce the numbers reported, you can use this data file (the original numbers were still a bit lower possibly because 1) we used dropout, but the same parameters are always applied to HiLAP too; 2) the reported performance is lower than best performance). Otherwise, you can just enjoy the free gains :)

**Comment by boxiaowave** [comment]
Hi，morningmoni

could you please provide the preload_data_rcv1_ptAll_onehot.pkl you used in your experiments? Since for bow_cnn, my results are also quite different from the results you report in your paper .
